src/Lykke.Service.OperationsHistory/Services/InMemoryCache/InMemoryCache.cs

```diff
@@ -1,172 +1,93 @@
+            if (!_storage.ContainsKey(clientId))
+            {
+                await Load(clientId);
+            }
         }
 
-        private async Task<IEnumerable<HistoryEntryResponse>> InternalGetAllAsync(string clientId, int top, int skip)
+        private async Task<CacheModel> Load(string clientId)
         {
-            var clientRecords = await GetRecordsByClient(clientId);
+            var records = await _repository.GetByClientIdAsync(clientId);
 
-            var pagedResult = clientRecords
-                .Skip(skip)
-                .Take(top)
-                .ToList();
+            if (!records.Any())
+                return null;
 
-            return Mapper.Map<IEnumerable<HistoryEntryResponse>>(pagedResult);
-        }
-
-        private async Task<IEnumerable<HistoryEntryResponse>> InternalGetAllAsync(string clientId, string assetId, string operationType,
-            int top, int skip)
-        {
-            var clientRecords = await GetRecordsByClient(clientId);
-
-            var pagedResult = clientRecords
-                .Where(r => r.Currency == assetId && r.OpType == operationType)
-                .Skip(skip)
-                .Take(top)
-                .ToList();
-
-            return Mapper.Map<IEnumerable<HistoryEntryResponse>>(pagedResult);
-        }
-
-        private async Task<IEnumerable<HistoryEntryResponse>> InternalGetAllByOpTypeAsync(string clientId, string operationType, int top, int skip)
-        {
-            var clientRecords = await GetRecordsByClient(clientId);
-
-            var pagedResult = clientRecords
-                .Where(r => r.OpType == operationType)
-                .Skip(skip)
-                .Take(top)
-                .ToList();
+            var cacheModel = new CacheModel
+            {
+                Records = new ConcurrentDictionary<string, IHistoryLogEntryEntity>(
+                    records
+                        .OrderByDescending(r => r.DateTime)
+                        .Select(x => new KeyValuePair<string, IHistoryLogEntryEntity>(x.Id, x)))
+            };
 
-            return Mapper.Map<IEnumerable<HistoryEntryResponse>>(pagedResult);
+            return _storage.AddOrUpdate(clientId, cacheModel, (key, oldValue) => cacheModel);
         }
 
-        private async Task<IEnumerable<HistoryEntryResponse>> InternalGetAllByAssetAsync(string clientId, string assetId, int top, int skip)
+        public async Task<IEnumerable<HistoryEntryResponse>> GetAsync(string clientId, string operationType, string assetId, int take, int skip)
         {
             var clientRecords = await GetRecordsByClient(clientId);
 
-            var pagedResult = clientRecords
-                .Where(r => r.Currency == assetId)
+            var result = clientRecords
+                .Where(r => string.IsNullOrWhiteSpace(operationType) || r.OpType == operationType)
+                .Where(r => string.IsNullOrWhiteSpace(assetId) || r.Currency == assetId)
```

Comment:
I would introduce temp variables before filtering instead of calls string.IsNullOrWhiteSpace().